spring-cloud-gcp-autoconfigure/src/main/java/com/google/cloud/spring/autoconfigure/trace/pubsub/brave/PubSubTracing.java

```diff
@@ -0,0 +1,316 @@
+	// We can't just skip clearing headers we use because we might inject B3 single, yet have stale B3
+	// multi, or visa versa.
+	void clearTraceIdHeaders(PubsubMessage.Builder message) {
+		for (String traceIDHeader : traceIdHeaders) {
+			message.removeAttributes(traceIDHeader);
+		}
+	}
+
+	/** When an upstream context was not present, lookup keys are unlikely added */
+	static void addTags(PubsubMessage.Builder message, SpanCustomizer result) {
+
+		// TODO: anything to do here?
+	}
+
+
+	PullResponse tracePullResponse(PullResponse delegate, String subscriptionName) {
+
+		if (delegate.getReceivedMessagesCount() == 0 || tracing.isNoop()) {
+			return delegate;
+		}
+
+		Span[] batchSpan = null;
+		if (singleRootSpanOnReceiveBatch) {
+			batchSpan = new Span[1];
+		}
+
+		PullResponse.Builder wrappedPullResponseBuilder = delegate.toBuilder();
+		for (int i = 0; i < delegate.getReceivedMessagesCount(); i++) {
+			wrappedPullResponseBuilder.setReceivedMessages(i, traceReceivedMessage(delegate.getReceivedMessages(i), subscriptionName, batchSpan));
+		}
+
+		return wrappedPullResponseBuilder.build();
+	}
+
+	private ReceivedMessage traceReceivedMessage(ReceivedMessage receivedMessage, String subscriptionName, Span[] batchSpan) {
+		PubsubMessage.Builder wrappedMessage = receivedMessage.getMessage().toBuilder();
+		postProcessMessageForConsuming(wrappedMessage, subscriptionName, batchSpan);
+		return receivedMessage.toBuilder()
+				.setMessage(wrappedMessage.build())
+				.build();
+	}
+
+	private void postProcessMessageForConsuming(PubsubMessage.Builder messageBuilder, String subscriptionName, Span[] batchSpan) {
+		long timestamp = 0;
+		PubSubConsumerRequest request = new PubSubConsumerRequest(messageBuilder, subscriptionName);
+		TraceContextOrSamplingFlags extracted =
+				extractAndClearTraceIdHeaders(consumerExtractor, request, messageBuilder);
+
+		// If we extracted neither a trace context, nor request-scoped data (extra),
+		// and sharing trace is enabled make or reuse a span for this topic
+		if (extracted.equals(emptyExtraction) && singleRootSpanOnReceiveBatch) {
+			Span span = batchSpan[0];
+			if (span == null) {
+				span = nextMessagingSpan(consumerSampler, request, extracted);
+				if (!span.isNoop()) {
+					setConsumerSpan(span, subscriptionName);
+
+					// incur timestamp overhead only once
+					if (timestamp == 0L) {
+						timestamp = tracing.clock(span.context()).currentTimeMicroseconds();
+					}
+					span.start(timestamp);
+				}
+				batchSpan[0] = span;
+			}
+			consumerInjector.inject(span.context(), request);
+		}
+		else { // we extracted request-scoped data, so cannot share a consumer span.
+			Span span = nextMessagingSpan(consumerSampler, request, extracted);
+			if (!span.isNoop()) {
+				setConsumerSpan(span, subscriptionName);
+
+				// incur timestamp overhead only once
+				if (timestamp == 0L) {
+					timestamp = tracing.clock(span.context()).currentTimeMicroseconds();
+				}
+				span.start(timestamp);
```

Comment:
this part can probably become a helper method; it's similar to above.

On the other hand, if we are putting these classes back to brave, it might help for maintenance there if the code looks similar to the rest of brave..

Reply:
Yep, for now trying to stay as close as possible to Brave code.